src/main/java/net/kyori/mu/collection/SyncMap.java

```diff
@@ -0,0 +1,261 @@
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * Iterations over a sync map are thread-safe, and the keys iterated over will not change for a single iteration
+   * attempt, however they may not necessarily reflect the state of the map at the time the iterator was created.
+   *
+   * <p>Performance note: if entries have been appended to the map, iterating over the entry set will automatically
+   * promote them to the read map.</p>
+   */
+  @NonNull
+  @Override
+  Set<Entry<K, V>> entrySet();
+
+  /**
+   * {@inheritDoc}
+   *
+   * This implementation is {@code O(n)} in nature due to the need to check for any expunged entries. Likewise, as
+   * with other concurrent collections, the value obtained by this method may be out of date by the time this method
+   * returns.
+   *
+   * @return the size of all the mappings contained in this map
+   */
+  @Override
+  int size();
+
+  /**
+   * {@inheritDoc}
+   *
+   * This method clears the map by resetting the internal state to a state similar to as if a new map had been created.
+   * If there are concurrent iterations in progress, they will reflect the state of the map prior to being cleared.
+   */
+  @Override
+  void clear();
+
+  /**
+   * The expunging value the backing map wraps for its values.
+   *
+   * @param <V> the backing value type
+   */
+  interface ExpungingValue<V> {
+    /**
+     * Returns the backing element, which may be {@code null} if it has been expunged.
+     *
+     * @return the backing element if it has not been expunged
+     */
+    @Nullable
+    V get();
+
+    /**
+     * Attempts to place the entry in the map if it is absent.
+     *
+     * @param value the value to place in the map
+     * @return a {@link Map.Entry} with false key and null value if the value was expunged, a true key and the
+     * previous value in the map otherwise
+     */
+    Map.Entry<Boolean, V> putIfAbsent(final @NonNull V value);
+
+    /**
+     * Returns {@code true} if this element has been expunged.
+     *
+     * @return whether or not this element has been expunged
+     */
+    boolean expunged();
+
+    /**
+     * Returns {@code true} if this element has a value (it is neither expunged nor {@code null}.
+     *
+     * @return whether or not this element has a value
+     */
+    boolean exists();
+
+    /**
+     * Sets the backing element.
+     *
+     * @param element the backing element
+     * @return the previous element stored, or {@code null} if the entry had been expunged
+     */
+    @Nullable
+    V set(final @NonNull V element);
+
+    /**
+     * Tries to replace the backing element, which can be set to {@code null}. This operation has no effect
+     * if the entry was expunged.
+     *
+     * @param expected the element to check for
+     * @param newElement the new element to be stored
+     * @return {@code true} if successful, {@code false} otherwise
+     */
+    boolean replace(final @NonNull Object expected, final @Nullable V newElement);
```

Comment:
Why `Object`?

Reply:
This method used by the map implementation only has the first parameter as an `Object` to avoid unnecessary type casting for the `Map#remove(Object,Object)` method. 

Used here: https://github.com/KyoriPowered/mu/pull/3/files#diff-ac650e0c21bf18317f891ae7a348b8f4R175

Although this can be casted if you still feel it is appropriate.